split.cpp

```diff
@@ -0,0 +1,116 @@
+#include <algorithm>
+#include <execution>
+#include <fstream>
+#include <iostream>
+#include <memory>
+#include <queue>
+#include <sstream>
+#include <string>
+#include <system_error>
+#include <unordered_map>
+#include <unordered_set>
+
+#include <llvm/ADT/SmallVector.h>
+#include <llvm/ADT/StringExtras.h>
+#include <llvm/ADT/StringRef.h>
+#include <llvm/Analysis/CallGraph.h>
+#include <llvm/Bitcode/BitcodeWriter.h>
+#include <llvm/IR/Argument.h>
+#include <llvm/IR/Constant.h>
+#include <llvm/IR/DerivedTypes.h>
+#include <llvm/IR/Function.h>
+#include <llvm/IR/GlobalValue.h>
+#include <llvm/IR/GlobalVariable.h>
+#include <llvm/IR/InstrTypes.h>
+#include <llvm/IR/Instruction.h>
+#include <llvm/IR/Instructions.h>
+#include <llvm/IR/Module.h>
+#include <llvm/IR/Use.h>
+#include <llvm/IR/InstVisitor.h>
+#include <llvm/IR/Operator.h>
+#include <llvm/IR/DerivedUser.h>
+#include <llvm/IRReader/IRReader.h>
+#include <llvm/Support/Casting.h>
+#include <llvm/Support/CommandLine.h>
+#include <llvm/Support/SourceMgr.h>
+#include <llvm/Support/raw_ostream.h>
+#include <llvm/Transforms/Utils/Cloning.h>
+#include <llvm/Transforms/Utils/ValueMapper.h>
+#include <llvm/Pass.h>
+
+#include <assert.h>
+#include <typeinfo>
+
+static llvm::LLVMContext context;
+llvm::SMDiagnostic err;
+
+llvm::cl::OptionCategory Splitter("Split");
+
+llvm::cl::opt<std::string> InputFilename(llvm::cl::Positional,
+                                         llvm::cl::desc("<input file>"));
+
+llvm::cl::opt<bool>
+    Verbose("v", llvm::cl::desc("Enable vrbose output."));
+
+struct MyPass : public llvm::InstVisitor<MyPass>  {
+  std::unordered_set<llvm::GlobalVariable *> globals;
+  
+  void visitInstruction(llvm::Instruction& instruction) {
+    std::stack<llvm::Value *> operands;
+      
+    for (uint32_t i = 0; i < instruction.getNumOperands(); i++) {
+      operands.push(instruction.getOperand(i));
+    }
+
+    while (operands.size() > 0) {
+      auto currentOperand = operands.top();
+      operands.pop();
+      
+      if (currentOperand == nullptr) {
+        //TODO: ignore null's for now.
```

Comment:
What would a null operand be?

Reply:
To be fair I checked and some instructions can take 2 or 3 operands, and the getNumOperands function returns the maximum they can have. This is why sometimes the operand can be null.

Reply:
Silly question: why does the number of operands mean they can return null? Should we be printing warnings here or does this happen quite often?